### ISSUE TITLE
R Language Formatter in Editor

### ISSUE THREAD

**Issue by gaocegege** [gaocegege/Processing.R#84]
Now there is no formatter to format the code, we need to import or implement one.

```java
  /**
   * @see processing.app.ui.Editor#createFormatter()
   */
  @Override
  public Formatter createFormatter() {
    return null;
  }
```

**Comment by jeremydouglass** [comment]
Perhaps use formatR?

-  https://yihui.name/formatr/
-  https://stackoverflow.com/questions/3017877/tool-to-auto-format-r-code

**Comment by gaocegege** [comment]
Yes, it may works, and I have some connections with the authors of formatR, so we have technical support :smile:

**Comment by jeremydouglass** [comment]
What are we currently using for checking R code formatting on the repository? *Ideally* the Example Processing.R sketches will both:

1. be unchanged when running the PDE R autoformatter
2. pass continuous integration code checks on the repository

**Comment by gaocegege** [comment]
We are using FormatRBear in coala, and its backend is formatR.

**Comment by gaocegege** [comment]
Yeah, I agree with you :)

**Comment by gaocegege** [comment]
Try to use formatR, which means if the user wanna to use the formatter, he/she needs R installed.

**Comment by jeremydouglass** [comment]
> if the user wanna to use the formatter, he/she needs R installed

As an alternative to the user installing R or to us including a full separate R install just for the PDE formatter...

Does formatR run in renjin?

http://packages.renjin.org/package/org.renjin.cran/formatR/1.2.1

Perhaps we could have the autoformatter call formatR through renjin, if it isn't too slow?

**Comment by gaocegege** [comment]
We could have a try but there are 5 tests failed.

**Comment by jeremydouglass** [comment]
I see -- then I think it is probably not worth trying.

**Comment by gaocegege** [comment]
There are two ways:

* formatR, we run formatR in a separate process and send the code to the process to do the check. it needs the users to install R and formatR package
* formatR in renjin, not sure if it works.
